lib/src/rules/prefer_is_empty.dart

```diff
@@ -0,0 +1,210 @@
+```
+''';
+
+class _LintCode extends LintCode {
+  static final registry = <String, LintCode>{};
+
+  factory _LintCode(String name, String message) => registry.putIfAbsent(
+      name + message, () => new _LintCode._(name, message));
+
+  _LintCode._(String name, String message) : super(name, message);
+}
+
+class PreferIsEmpty extends LintRule {
+  PreferIsEmpty()
+      : super(
+            name: 'prefer_is_empty',
+            description: desc,
+            details: details,
+            group: Group.style);
+
+  @override
+  AstVisitor getVisitor() => new Visitor(this);
+
+  void reportLintWithDescription(AstNode node, String description) {
+    if (node != null) {
+      reporter.reportErrorForNode(new _LintCode(name, description), node, []);
+    }
+  }
+}
+
+class Visitor extends SimpleAstVisitor {
+  final PreferIsEmpty rule;
+  Visitor(this.rule);
+
+  @override
+  visitSimpleIdentifier(SimpleIdentifier identifier) {
+    AstNode lengthAccess;
+    SimpleIdentifier lengthIdentifier;
+    InterfaceType type;
+
+    AstNode parent = identifier.parent;
+    if (parent is PropertyAccess) {
+      lengthIdentifier = parent.propertyName;
+      lengthAccess = parent;
+      if (parent.target?.bestType is! InterfaceType) {
+        return;
+      }
+      type = parent.target?.bestType;
+    } else if (parent is PrefixedIdentifier) {
+      lengthIdentifier = parent.identifier;
+      lengthAccess = parent;
+      if (parent.prefix.bestType is! InterfaceType) {
+        return;
+      }
+      type = parent.prefix.bestType;
+    }
+
+    if (lengthIdentifier == null) {
+      return;
+    }
+
+    // Should be "length".
+    Element propertyElement = lengthIdentifier.bestElement;
+    if (propertyElement?.name != 'length') {
+      return;
+    }
+
+    // Should be subtype of Iterable, Map or String.
+    AnalysisContext context = propertyElement.context;
+    TypeProvider typeProvider = context.typeProvider;
+    TypeSystem typeSystem = context.typeSystem;
+
+    if (typeSystem.mostSpecificTypeArgument(type, typeProvider.iterableType) ==
+            null &&
+        typeSystem.mostSpecificTypeArgument(type, typeProvider.mapType) ==
+            null &&
+        !type.element.type.isSubtypeOf(typeProvider.stringType)) {
+      return;
+    }
+
+    AstNode search = lengthAccess;
+    while (
+        search != null && search is Expression && search is! BinaryExpression) {
+      search = search.parent;
+    }
+
+    if (search is! BinaryExpression) {
+      return;
+    }
+    BinaryExpression binaryExpression = search;
+
+    Token operator = binaryExpression.operator;
+    DeclaredVariables declaredVariables = context.declaredVariables;
+
+    // Comparing constants with length.
+
+    DartObjectImpl rightValue = binaryExpression.rightOperand.accept(new ConstantVisitor(
+        new ConstantEvaluationEngine(typeProvider, declaredVariables,
+            typeSystem: typeSystem),
+        rule.reporter));
```

Comment:
I missed this the first time, but I think passing in `rule.reporter` will cause any errors produced by the ConstantVisitor to be reported twice. I think you want `AnalysisErrorListener.NULL_LISTENER`.